### PR TITLE
Mentors vol. 2

### PR DESCRIPTION
**Pull request by 1fbff5f83b23d39d38b1dfcb4cac8d9b**
Pretty much a reboot of #1854. From the PR:

From what I understand Mentors are a lower-class admin that are coded entirely differently.
Mentors are appointed by the council. They work off of a notepad file.
Mentors also have mentorSAY, for chatting like little girls, and SHOW mentor memos (those are set by admins).
Mentors also have the ability to follow players. This'll set their entire perspective on that player.

Players will have the ability to use mentorhelps in order to communicate with mentors. Mentors will also have the ability to PM players, and communicate back.

And so they get the recognition they deserve their OOC color is also different. I might as well throw in an OOC tag.

Thanks goes out to the hippiestation team.

Ports:

HippieStation/HippieStation13#22
HippieStation/HippieStation13#73
HippieStation/HippieStation13#170
Changelog

:cl: McDonald072
rscadd: Ported Mentors from HippieStation.
/:cl:

### CODE REVIEW

**Comment by ShadowDeath6** [comment]
put a name in the changelog so people don't see "1fbff5f83b23d39d38b1dfcb4cac8d9b" in game, please

**Comment by drderp3635** [comment]
1fbff5f83b23d39d38b1dfcb4cac8d9b is a beautiful name, don't let them bully you

**Comment by Super3222** [comment]
You should just sign the changelog under your username. RN it looks like Hippiestation is doing the porting which is confusing. Either way nice job.

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
Doesn't feel right to attribute this to myself when I essentially copy-pasted your work to be honest

**Comment by Alexandria404** [comment]
Where are the logs for their actions and what is logged, and can it be in it´s own file?

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
It goes in the server log. The following are logged: following, unfollowing, people mentorhelping and replies to those, and mentorsay. The logs could be stored in a separate file if need be (but pls dont ask for that i dunno how to)

**Comment by Amelia0010** [comment]
pleeeeeeaaaaasee put it in the admin log, rather than the regular log

**Comment by Xantam-13** [comment]
Forgive me if I'm being silly, but, I'm assuming this will now include 2 ticket systems, one for ahelps and one for mentor helps? Both will be separate and not buggy together? Or will ahelps use tickets and mentor helps use the old style PM'ing without tickets?

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
Mentorhelp will work via oldschool PMs

File: code/modules/client/verbs/ooc.dm [hunk 1]
```diff
@@ -94,6 +94,10 @@
 					C << "<font color='[normal_ooc_colour]'><span class='ooc'><span class='prefix'>OOC:</span> <EM>[holder.fakekey ? holder.fakekey : key]:</EM> <span class='message'>[msg]</span></span></font>"
 
 			else if(!(key in C.prefs.ignoring))
+				var/ooc_color = normal_ooc_colour
+				if(check_mentor())
+					ooc_color = "#FF3E96"
+				C << "<font color='[ooc_color]'><span class='ooc'><span class='prefix'>OOC:</span> <EM>[keyname]:</EM> <span class='message'>[msg]</span></span></font>"
```
Comment (ghost) [inline code/modules/client/verbs/ooc.dm]: to_chat

File: code/modules/client/verbs/who.dm [hunk 2]
```diff
@@ -47,11 +47,19 @@
...
+			var/mentor = mentor_datums[C.ckey]
+			if(mentor)
+				Lines += "<font color='#0033CC'>\t[C.key]</font> - Mentor"
 
 	for(var/line in sortList(Lines))
 		msg += "[line]\n"
 
-	msg += "<b>Total Players: [length(Lines)]</b>"
+	msg += "<b>Total Players: [player_number]</b>"
 	src << msg
```
Comment (ghost) [inline code/modules/client/verbs/who.dm]: to_chat

File: code/modules/hippie/mentor/follow.dm [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		mentor.following = M
+	else
+		holder.following = M
+
+	usr.reset_perspective(M)
+	usr.client.click_intercept = new /datum/mentor_click_interceptor
+	usr.client.mob.notransform = TRUE
+	src.verbs += /client/proc/mentor_unfollow
+
+	admins << "<span class='mentor'><span class='prefix'>MENTOR:</span> <EM>[key_name(usr)]</EM> is now following <EM>[key_name(M)]</span>"
```
Comment (ghost) [inline code/modules/hippie/mentor/follow.dm]: to_chat

File: code/modules/hippie/mentor/follow.dm [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	var/following = null
+	if(!holder)
+		var/datum/mentors/mentor = mentor_datums[usr.client.ckey]
+		following = mentor.following
+	else
+		following = holder.following
+
+
+	admins << "<span class='mentor'><span class='prefix'>MENTOR:</span> <EM>[key_name(usr)]</EM> is no longer following <EM>[key_name(following)]</span>"
+	usr << "<span class='info'>You are no longer following [following].</span>"
```
Comment (ghost) [inline code/modules/hippie/mentor/follow.dm]: to_chat

Comment (ghost) [inline code/modules/hippie/mentor/follow.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentor_memo.dm [hunk 5]
```diff
@@ -0,0 +1,139 @@
+/client/proc/mentor_memo()
+	set name = "Mentor Memos"
+	set category = "Server"
+	if(!check_rights(0))	return
+	if(!dbcon.IsConnected())
+		src << "<span class='danger'>Failed to establish database connection.</span>"
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentor_memo.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentor_memo.dm [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	if(!memotask)
+		return
+	mentor_memo_output(memotask)
+
+/client/proc/show_mentor_memo()
+	set name = "Show Memos"
+	set category = "Mentor"
+	if(!check_mentor())	return
+	if(!dbcon.IsConnected())
+		src << "<span class='danger'>Failed to establish database connection.</span>"
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentor_memo.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentor_memo.dm [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+		return
+	mentor_memo_output("Show")
+
+/client/proc/mentor_memo_output(task)
+	if(!task)
+		return
+	if(!dbcon.IsConnected())
+		src << "<span class='danger'>Failed to establish database connection.</span>"
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentor_memo.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentor_memo.dm [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	var/sql_ckey = sanitizeSQL(src.ckey)
+	switch(task)
+		if("Write")
+			var/DBQuery/query_memocheck = dbcon.NewQuery("SELECT ckey FROM [format_table_name("mentor_memo")] WHERE ckey = '[sql_ckey]'")
+			if(!query_memocheck.Execute())
+				var/err = query_memocheck.ErrorMsg()
+				log_game("SQL ERROR obtaining ckey from memo table. Error : \[[err]\]\n")
+				return
+			if(query_memocheck.NextRow())
+				src << "You already have set a memo."
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentor_memo.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentor_memo.dm [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+			if(!query_memolist.Execute())
+				var/err = query_memolist.ErrorMsg()
+				log_game("SQL ERROR obtaining ckey from memo table. Error : \[[err]\]\n")
+				return
+			var/list/memolist = list()
+			while(query_memolist.NextRow())
+				var/lkey = query_memolist.item[1]
+				memolist += "[lkey]"
+			if(!memolist.len)
+				src << "No memos found in database."
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentor_memo.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentor_memo.dm [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+				var/ckey = query_memoshow.item[1]
+				var/memotext = query_memoshow.item[2]
+				var/timestamp = query_memoshow.item[3]
+				var/last_editor = query_memoshow.item[4]
+				output += "<span class='memo'>Mentor memo by <span class='prefix'>[ckey]</span> on [timestamp]"
+				if(last_editor)
+					output += "<br><span class='memoedit'>Last edit by [last_editor] <A href='?_src_=holder;mentormemoeditlist=[ckey]'>(Click here to see edit log)</A></span>"
+				output += "<br>[memotext]</span><br>"
+			if(!output)
+				src << "No memos found in database."
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentor_memo.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentor_memo.dm [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+				return
+			src << output
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentor_memo.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentor_memo.dm [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+			if(!query_memodellist.Execute())
+				var/err = query_memodellist.ErrorMsg()
+				log_game("SQL ERROR obtaining ckey from memo table. Error : \[[err]\]\n")
+				return
+			var/list/memolist = list()
+			while(query_memodellist.NextRow())
+				var/ckey = query_memodellist.item[1]
+				memolist += "[ckey]"
+			if(!memolist.len)
+				src << "No memos found in database."
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentor_memo.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentorhelp.dm [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+	//spawn(300)
+	//	src.verbs += /client/verb/mentorhelp	// 30 second cool-down for mentorhelp
+
+	//clean the input msg
+	if(!msg)	return
+	msg = sanitize(copytext(msg,1,MAX_MESSAGE_LEN))
+	if(!msg)	return
+	if(!mob)	return						//this doesn't happen
+	if(src.prefs.muted & MUTE_MENTORHELP)
+		src << "<font color='red'>You are unable to use mentorhelp (muted).</font>"
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentorhelp.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentorhelp.dm [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+	var/mentor_msg = "<span class='mentornotice'><b><font color='#3280ff'>MENTORHELP:</b> <b>[key_name_mentor(src, 1, 0, 1, show_char)]</b>:</font> [msg]</span>"
+	log_mentor("MENTORHELP: [key_name_mentor(src, 0, 0, 0, 0)]: [msg]")
+
+	for(var/client/X in mentors)
+		X << 'sound/items/bikehorn.ogg'
+		X << mentor_msg
+
+	for(var/client/A in admins)
+		A << 'sound/items/bikehorn.ogg'
+		A << mentor_msg
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentorhelp.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentorhelp.dm [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+	if(src.prefs.muted & MUTE_MENTORHELP)
+		src << "<font color='red'>You are unable to use mentorhelp (muted).</font>"
+		return
+	var/show_char = config.mentors_mobname_only
+	var/mentor_msg = "<span class='mentornotice'><b><font color='#3280ff'>MENTORHELP:</b> <b>[key_name_mentor(src, 1, 0, 1, show_char)]</b>:</font> [msg]</span>"
+	log_mentor("MENTORHELP: [key_name_mentor(src, 0, 0, 0, 0)]: [msg]")
+
+	for(var/client/X in mentors)
+		X << 'sound/items/bikehorn.ogg'
+		X << mentor_msg
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentorhelp.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentorpm.dm [hunk 16]
```diff
@@ -0,0 +1,74 @@
+//shows a list of clients we could send PMs to, then forwards our choice to cmd_Mentor_pm
+/client/proc/cmd_mentor_pm_panel()
+	set category = "Mentor"
+	set name = "Mentor PM"
+	if(!holder)
+		src << "<font color='red'>Error: Mentor-PM-Panel: Only Mentors may use this command.</font>"
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentorpm.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentorpm.dm [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+			src << "<font color='green'>Mentor PM to-<b>[key_name_mentor(C, C, 1, 0, 0)]</b>: [msg]</font>"
+
+		else		//recipient is an mentor but sender is not
+			C << "<font color='#3280ff'>Reply PM from-<b>[key_name_mentor(src, C, 1, 0, show_char)]</b>: [msg]</font>"
+			src << "<font color='green'>Mentor PM to-<b>[key_name_mentor(C, C, 1, 0, 0)]</b>: [msg]</font>"
+
+	else
+		if(check_mentor())	//sender is an mentor but recipient is not.
+			C << "<font color='#3280ff'>Mentor PM from-<b>[key_name_mentor(src, C, 1, 0, 0)]</b>: [msg]</font>"
+			src << "<font color='green'>Mentor PM to-<b>[key_name_mentor(C, C, 1, 0, show_char)]</b>: [msg]</font>"
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentorpm.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentorpm.dm [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+	C << 'sound/items/bikehorn.ogg'
+	var/show_char = config.mentors_mobname_only
+	if(check_mentor_other(C))
+		if(check_mentor())	//both are mentors
+			C << "<font color='#3280ff'>Mentor PM from-<b>[key_name_mentor(src, C, 1, 0, 0)]</b>: [msg]</font>"
+			src << "<font color='green'>Mentor PM to-<b>[key_name_mentor(C, C, 1, 0, 0)]</b>: [msg]</font>"
+
+		else		//recipient is an mentor but sender is not
+			C << "<font color='#3280ff'>Reply PM from-<b>[key_name_mentor(src, C, 1, 0, show_char)]</b>: [msg]</font>"
+			src << "<font color='green'>Mentor PM to-<b>[key_name_mentor(C, C, 1, 0, 0)]</b>: [msg]</font>"
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentorpm.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentorpm.dm [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+
+	log_mentor("Mentor PM: [key_name(src)]->[key_name(C)]: [msg]")
+
+	msg = emoji_parse(msg)
+	C << 'sound/items/bikehorn.ogg'
+	var/show_char = config.mentors_mobname_only
+	if(check_mentor_other(C))
+		if(check_mentor())	//both are mentors
+			C << "<font color='#3280ff'>Mentor PM from-<b>[key_name_mentor(src, C, 1, 0, 0)]</b>: [msg]</font>"
+			src << "<font color='green'>Mentor PM to-<b>[key_name_mentor(C, C, 1, 0, 0)]</b>: [msg]</font>"
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentorpm.dm]: to_chat

File: code/modules/hippie/mentor/verbs/mentorsay.dm [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+	if(!msg)	return
+
+	msg = emoji_parse(msg)
+	log_mentor("MSAY: [key_name(src)] : [msg]")
+
+
+	if(check_rights(R_ADMIN,0))
+		msg = "<span class='mentoradmin'><span class='prefix'><font color='#3280ff'>MENTOR:</font></span> <EM>[key_name(src, 0, 0)]</EM>: <span class='message'>[msg]</span></span>"
+		mentors << msg
+		admins << msg
```
Comment (ghost) [inline code/modules/hippie/mentor/verbs/mentorsay.dm]: to_chat x4 here

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
OK Council has agreed to this as soon as it's observer only.  I was thinking dementor them when they get a job and then not allowing rementor until they are observer again.  Someone else probably has a better solution.

**Comment by Koffheatalt** [comment]
isn't it better to use 
if(!isliving(user))

**Comment by ghost** [comment]
Yeah probably best to just put that literally every place that mentors have abilities.

**Comment by Koffheatalt** [comment]
fair point it would create a lot of lagg

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
what would create lag

**Comment by Koffheatalt** [comment]
constantly checking if(!isliving(target)) for each mentor

**Comment by Amelia0010** [comment]
You'd only need to check it everytime someone mentorhelps or uses mentorsay or something similar. And doing a few if statements really wouldn't lag that much.

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
ayy lmao doesnt work

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
this should work but i cant test rihgt now

**Comment by Koffheatalt** [comment]
did you test it yet

**Comment by 1fbff5f83b23d39d38b1dfcb4cac8d9b** [comment]
how did you get into my git mainframe and when is this getting testmerged

**Pull request closed without merging**